keps/sig-api-machinery/2340-Consistent-reads-from-cache/README.md

```diff
@@ -125,29 +124,91 @@ serves the resourceVersion="0" list requests from reflectors today.
+section.
+
+A single request is not enough to ensure that the wait on resource version
+terminates. If etcd receives a burst of events, the watch might be in the middle
+of processing a large batch of events when the watch cache requests progress.
+The resulting progress notification will return a resource version earlier than
+requested, which could cause an infinite wait if there are no follow-up events.
+
+To rely on progress notification for freshness, the watch cache will
+periodically (every 100ms) request progress until it reaches the requested
+revision or the request times out.
+
+#### The algorithm
 
 When a consistent LIST request is received and the watch cache is enabled:
 
 - Get the current revision from etcd for the resource type being served.
-  Use the [getCurrentResourceVersionFromStorage] added as part of [Watch-List KEP].
-- If the cache already has the current revision, serve the request from cache. If not,
-  - Send a `WatchProgressRequest` to etcd on the watch channel that the watch cache is consuming.
-- Use the existing `waitUntilFreshAndBlock` function in the watch cache to wait briefly for the watch to catch up to the current revision.
-- If the block times out, the request will result in rejection. (see "What if the watch cache is stale?" section for details)
+- Check if the watch cache already has the current revision, if not:
+  - Add a "waiting read" and notify the goroutine running in background.
+  - Wait for watch cache to catch up. If wait times out, reject the request.
+    (see "What if the watch cache is stale?" section for details)
+  - Remove a "waiting read"
+- Serve the request from watch cache
+
+In background, run a dedicated goroutine that will:
+- Wait for changes to "waiting read" count.
+- Repeat as long there is at least one "waiting read":
+  - Send `WatchProgressRequest` request to etcd.
+  - Wait 100ms
 
 Consistent GET requests will continue to be served directly from etcd. We will
 only serve consistent LIST requests from cache.
 
-[getCurrentResourceVersionFromStorage]: https://github.com/kubernetes/kubernetes/blob/3f247e59edfd4083242ad7271d076a38291760ff/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L1246-L1278
 [Watch-List KEP]: /keps/sig-api-machinery/3157-watch-list
 
+### Bug in etcd progress notification
+
+Only recently community discovered a bug [etcd-io/etcd#15220] for requesting
+progress notification. This bug causes a race between sending an event and
+progress notification with the same revision. Normally we would expect to event
+to always come first, however due to the bug, progress notification might be
+sent first. Serving a consistent lists in that situation could result in
+returning too early and missing the event. The bug was only fixed in v3.4.25 and
+v3.5.8 (both 9 months old), which is too fresh for Kubernetes to not handle.
+
+Our tests have shown that with the bug manual progress notification cannot be
```

Comment:
Clarifying because it came up for me.  This etcd bug results in a client getting incorrect data from the request in a way that the client is unaware of and the server is unaware of.  In other words, a silent data corruption bug that cannot be automatically detect prior to acting upon the corrupted data.

If this were a GA feature, we'd have no choice but to leave it enabled, but because it is a beta feature, we're able to conditionally enable in a way that gives us confidence enabling for beta (data is correct) without risking undetectable failures due to slightly old levels.

Reply:
Updated the section.